File: src/main/java/info/kfgodel/function/MemoizedSupplier.java
package info.kfgodel.function;

import java.util.function.Supplier;

/**
 * This type represents a supplier that keeps the first value it produces and returns that value on subsequent calls.<br>
 *   This class allows reusing created objects, while keeping a simple interface for creating them
 * Date: 04/05/19 - 18:21
 */
public class MemoizedSupplier<T> implements Supplier<T> {

  private Supplier<T> delegate;
  private T memoizedValue;

  @Override
  public T get() {
    if(delegate != null){
      this.memoizedValue = delegate.get();
      this.delegate = null; // Discard the original supplier after use
    }
    return memoizedValue ;
  }

  public static <T>  MemoizedSupplier<T> create(Supplier<T> onlyInstanceSupplier) {
    MemoizedSupplier memoizedSupplier = new MemoizedSupplier();
    memoizedSupplier.delegate = onlyInstanceSupplier;
    return memoizedSupplier;
  }

  /**
   * Friendly alias of {@link #create(Supplier)} to be imported in other classes.<br>
   *   It creates a supplier that will hold a reference to the first value produced by the
   *   given supplier and keep returning that when called
   * @param other The supplier whose first value is to be memoized
   * @param <T> The type of returned value
   * @return The memorized supplier
   */
  public static <T> Supplier<T> memoized(Supplier<T> other){
    return create(other);
  }

}

File: src/main/kotlin/info/kfgodel/function/MemoizedSupplier.kt
package info.kfgodel.function

import java.util.function.Supplier

/**
 * This type represents a supplier that keeps the first value it produces and returns that value on subsequent calls.<br></br>
 * This class allows reusing created objects, while keeping a simple interface for creating them
 * Date: 04/05/19 - 18:21
 */
class MemoizedSupplier<T> : Supplier<T> {
  private var delegate: Supplier<T>? = null
  private var memoizedValue: T? = null
  override fun get(): T {
    if (delegate != null) {
      memoizedValue = delegate!!.get()
      delegate = null // Discard the original supplier after use
    }
    return memoizedValue!!
  }

  companion object {
    fun <T> create(onlyInstanceSupplier: Supplier<T>?): MemoizedSupplier<T> {
      val memoizedSupplier: MemoizedSupplier<T> = MemoizedSupplier<T>()
      memoizedSupplier.delegate = onlyInstanceSupplier
      return memoizedSupplier
    }

    /**
     * Friendly alias of [.create] to be imported in other classes.<br></br>
     * It creates a supplier that will hold a reference to the first value produced by the
     * given supplier and keep returning that when called
     * @param other The supplier whose first value is to be memoized
     * @param <T> The type of returned value
     * @return The memorized supplier
    </T> */
    fun <T> memoized(other: Supplier<T>?): Supplier<T> {
      return create(other)
    }
  }
}
